### PR TITLE
feat: add support for podman credential location

### LINKED ISSUE #216: [BUG] oc-mirror should support podman credentials location for registry authentication
# Version

```console
$ oc-mirror version
v0.1.0-alpha.3
```

# What happened?

I used `podman login` to set up my registry credentials and `oc-mirror` does not look in the location where those are saved.

# What you expected to happen?

`oc-mirror` checks `${XDG_RUNTIME_DIR}/containers/auth.json` for registry credentials

# How to reproduce it (as minimally and precisely as possible)?

Set up a test registry protected by basic auth
```console
$ podman login localhost:5000
$ oc-mirror --from mirror_seq1_000000.tar docker://localhost:5000
```

# Anything else we need to know?

This is marked as an RFE because the docs explicitly say to use `~.docker/config.json` and using this location to store registry credentials works.

The podman location is not checked because two upstream libraries use a registry client that does not check this location.
- `go-containerregsitry authn.DefaultKeychain` includes this in the main branch and will in the next release
- `operator-registry` uses a docker config library that hard codes the name of the auth file to `config.json`. Only the directory of that file can be configured.

# References

### PR DESCRIPTION
**Pull request by jpower432**
Closes #216

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

# Description

This PR adds the ability for `oc-mirror` to discover credentials at either well-known `docker` locations or `podman` locations. Defaults to `docker`. This PR also removes unused code pertaining to pull-secrets provided in the configuration spec.

*Note to the reviewer - this is still pointing to a fork of `operator-registry` to maintain the ability to merge declarative configs temporarily.*

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

- Manually tested

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by afflom** [comment]
@dinhxuanvu I've tested this and verified that the oc mirroring imports, operator registry imports, and go-containerregistry imports are respecting the podman credentials location (`$XDG_RUNTIME_DIR/containers/auth.json`). Sending to you for the lgtm.

File: pkg/image/testdata/v2/single_manifest/manifests/oc-mirror [hunk 1]
```diff
@@ -0,0 +1 @@
+sha256:d31c6ea5c50be93d6eb94d2b508f0208e84a308c011c6454ebf291d48b37df19
```
Comment (dinhxuanvu) [inline pkg/image/testdata/v2/single_manifest/manifests/oc-mirror:1]: I'm not sure why an image SHA is here. Is this for testing reason?

File: README.md [hunk 2]
```diff
@@ -118,8 +118,7 @@ storageConfig:
   ```
 For configuration and options, see the [expanded overview](./docs/overview.md) and [usage](./docs/usage.md) docs.
 
-<sup>1</sup> For this example, the `create` and `publish` steps are run on the same machine. Therefore your `~/.docker/config.json`
-should contain auth config for both release/catalog source images _and_ your mirror registry.
+<sup>1</sup> For this example, the `create` and `publish` steps are run on the same machine. Therefore your `~/.docker/config.json` or `${XDG_RUNTIME_DIR}/containers/auth. json` should contain auth config for both release/catalog source images _and_ your mirror registry.
```
Comment (dinhxuanvu) [inline README.md:121]: There is an unintentional white space between `auth.` and `json` I think. Or maybe this is intentional?

File: docs/usage.md [hunk 3]
```diff
@@ -24,9 +24,9 @@
...
 
 ## Prerequisites
-> **WARNING**: Depending on the configuration file used and the periodicity between running `oc-mirror`, this process may download multiple hundreds of gigabytes of data, though differential updates should usually result in significantly smaller imagesets.
+> **WARNING**: Depending on the configuration file used and the periodicity between running `oc-mirror`, this process may download multiple hundreds of gigabytes of data, though differential updates should usually result in significantly smaller Imagesets.
 ### Authentication: 
-oc-mirror currently retrieves registry credentials from `~/.docker/config.json`. Make sure that your [Red Hat OpenShift Pull Secret](https://console.redhat.com/openshift/install/pull-secret) and any other needed registry credentials are populated in `~/.docker/config.json`
+oc-mirror currently retrieves registry credentials from `~/.docker/config.json` or `${XDG_RUNTIME_DIR}/containers/auth. json`. Make sure that your [Red Hat OpenShift Pull Secret](https://console.redhat.com/openshift/install/pull-secret) and any other needed registry credentials are populated in the credentials file.
```
Comment (dinhxuanvu) [inline docs/usage.md:29]: Same as above comment on the white space.

File: pkg/metadata/store.go [hunk 4]
```diff
@@ -44,7 +44,7 @@ func UpdateMetadata(ctx context.Context, backend storage.Backend, meta *v1alpha1
 func resolveOperatorMetadata(ctx context.Context, operator v1alpha1.Operator, backend storage.Backend, insecure bool) (operatorMeta v1alpha1.OperatorMetadata, err error) {
 	operatorMeta.Catalog = operator.Catalog
 
-	resolver, err := containerdregistry.NewResolver(operator.PullSecret, insecure, nil)
+	resolver, err := containerdregistry.NewResolver("", insecure, nil)
```
Comment (dinhxuanvu) [inline pkg/metadata/store.go:47]: From all the changes in this PR, I gather that if users want to include any secrets (such as pull secret), they will need to include that as a part of Docker/Podman setting. The option of providing discrete secrets outside of the daemons is essentially gone. Would that be a problem with someone wants to provide pull secrets with the set of images but has no access to daemons (such as lacking of admin access on the cluster)? Or we are saying the mirror operation is privileged by default?

**Review by dinhxuanvu — COMMENTED** [review]
Looking good. Just some comments/questions.

File: README.md (hunk 2 above)
Comment (jpower432) [reply README.md:121]: Copy and paste error :) Thanks for catching it.

File: docs/usage.md (hunk 3 above)
Comment (jpower432) [reply docs/usage.md:29]: Same

File: pkg/image/testdata/v2/single_manifest/manifests/oc-mirror (hunk 1 above)
Comment (jpower432) [reply pkg/image/testdata/v2/single_manifest/manifests/oc-mirror:1]: Might need to add this to `.gitignore`. Getting generated by the unit test.

File: pkg/metadata/store.go (hunk 4 above)
Comment (jpower432) [reply pkg/metadata/store.go:47]: You bring up a good point here. I would say that the user can at least construct their pull secret and put it in the expected location without needing privileged access. Our original intention with adding the pull secret to the spec was to have it directly in the config, but it quickly made the config unreadable. Eventually, we could add a flag to allow the user to specify a custom location for their pull secret. Still, we would either have to move it into an expected location or PR some of our dependencies to support this. WDYT?

Comment (dinhxuanvu) [reply pkg/metadata/store.go:47]: That sounds reasonable.

**Review by dinhxuanvu — COMMENTED** [review]
/lgtm

**Pull request merged**